agent/connectionuptime/service.go

```diff
@@ -0,0 +1,168 @@
+	}
+}
+
+// SetWindowPeriod updates window period
+func (c *Service) SetWindowPeriod(windowPeriod time.Duration) {
+	c.mx.Lock()
+	defer c.mx.Unlock()
+
+	c.windowPeriod = windowPeriod
+}
+
+// RegisterConnectionStatus adds connection event
+func (c *Service) RegisterConnectionStatus(timestamp time.Time, connected bool) {
+	c.mx.Lock()
+	defer c.mx.Unlock()
+	c.addEvent(timestamp, connected)
+}
+
+func (c *Service) addEvent(timestamp time.Time, connected bool) {
+	if len(c.events) == 0 || c.events[len(c.events)-1].Connected != connected {
+		c.events = append(c.events, connectionEvent{
+			Timestamp: timestamp,
+			Connected: connected,
+		})
+	}
+}
+
+func (c *Service) deleteOldEvents() {
+	c.mx.Lock()
+	defer c.mx.Unlock()
+
+	if len(c.events) == 0 {
+		return
+	}
+
+	// Move first elements which are already expired to the start of the slice
+	// in order to not lose information about previous state of connection.
+	// The latest expired element in the slice will be the first one to calculate
+	// uptime correctly during set up window time
+	expired := time.Now().Add(-c.windowPeriod)
+	for i := len(c.events) - 1; i >= 0; i-- {
+		if c.events[i].Timestamp.Before(expired) {
+			c.events[i].Timestamp = expired.Add(time.Second)
+			c.removeFirstElementsUntilIndex(i)
+
+			return
+		}
+	}
+}
+
+func (c *Service) removeFirstElementsUntilIndex(i int) {
+	c.events = append(c.events[:0], c.events[i:]...)
+}
+
+// RunCleanupGoroutine starts goroutine which removes already expired connection events.
+// Expired event means that it was created more than `windowPeriod` time ago.
+func (c *Service) RunCleanupGoroutine(ctx context.Context) {
+	go func() {
+		ticker := time.NewTicker(periodForRunningDeletingOldEvents)
+		for {
+			select {
+			case <-ticker.C:
+				c.l.Debug("Called delete old events")
+				c.deleteOldEvents()
+			case <-ctx.Done():
+				c.l.Debug("Done")
+				return
+			}
+		}
+	}()
+}
+
+// GetConnectedUpTimeSince calculates the connection up time between agent and server
+// based on the stored connection events.
+//
+// In the connection event set we store only when connection status was changed
+// (was it connected or not) in the next format:
+// {<timestamp, is_connected>, <timestamp, is_connected>, ...}
+//
+// For example:
+// {<'2022-01-01 15:00:00', true>, <'2022-01-01 15:20:00', false>, <'2022-01-01 15:20:10', true>}
+//
+// GetConnectionUpTime returns the percentage of connection uptime during
+// set period of time (by default it's 24 hours).
+// Method will calculate connected time as interval between connected and disconneced events
+//
+// Here is example how it works.
+// When we have such set of events in connection set `f1 s1 f2`
+// where f1 - first event of failed connection
+//       s1 - first event of successful connection
+//       f2 - second event of failed connection
+//
+// method will return result using next formula `time_between(s1, f2)/time_between(f1, now)*100`
```

Comment:
This formula is correct. First version of implementation this was used that formula which is in the ticket, but it was decided to change algorithm because previous one ` num_of_successfull_requests_to_server/num_of_total_requests_to_server*100` is not accurate.

Reply:
sounds good, thank you. I suggest we update the ticket as well.